pkg/controllers/provisioning/provisioner.go

```diff
@@ -104,27 +103,33 @@ func (p *Provisioner) provision(ctx context.Context) error {
 	if err != nil {
 		return fmt.Errorf("getting instance types, %w", err)
 	}
+
 	// Separate pods by scheduling constraints
 	nodes, err := p.scheduler.Solve(ctx, p.Provisioner, instanceTypes, pods)
 	if err != nil {
 		return fmt.Errorf("solving scheduling constraints, %w", err)
 	}
+	// we've got the set of nodes that we need to launch, but want to take advantage of any capability we have to
```

Comment:
Removing the aggregations would drastically reduce the performance of node launches. You won't see a drop off until you hit the rate limit, but at that point launches would come to a crawl. The rate limit we would hit first is EC2 Fleet for the AWS CP. EC2 has two rate limits when launching instances, one is the control-plane rate limiting on the CreateFleet call and the other is a resource limit on how many instances can be launched. When we aggregate the instances launched, we can increase our rate since we're doing less EC2 control-plane calls.  

IMO the ugliness is worth the performance improvement and since this is handled in core, CPs have an easier time. If we remove or shifted this to the CP side, I think most CPs would want to batch the launches to avoid API rate limiting for large scale-ups anyways and would result in a lot more ugliness in more places.

Reply:
I'm not sure I agree. What are the limits? Can they be raised? Most cloud providers don't have a fleet like API, so this is unnecessary complexity to them. Further, I think we could factor out a generic batcher that could solve this for all providers as a library if we needed it. 

Let's sync on slack.